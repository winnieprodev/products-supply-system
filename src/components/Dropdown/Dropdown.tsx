import { useTimeout } from "hooks";
import React, { useCallback, useEffect, useState } from "react";
import { useLocation } from "react-router";

import * as Styled from "./dropdown.styled";

interface OwnProps {
  anchor: React.ReactElement;
  closeOnRouteSwitch?: boolean;
}

enum State {
  closed = "closed",
  open = "open",
  closing = "closing",
}

const Dropdown: React.FC<React.PropsWithChildren<OwnProps>> = ({
  anchor,
  closeOnRouteSwitch,
  children,
}) => {
  const [state, setState] = useState<State>(State.closed);
  const location = useLocation();

  useEffect(() => {
    if (closeOnRouteSwitch && location && state === State.open) {
      setState(State.closing);
    }
  }, [closeOnRouteSwitch, location]);

  const toggleState = useCallback(
    () =>
      setState((prev) => (prev === State.open ? State.closing : State.open)),
    [],
  );

  const cleanup = useCallback(() => {
    setState(State.closed);
  }, []);

  useTimeout(cleanup, Styled.CLOSE_ANIMATION_TIME, state === State.closing);

  return (
    <Styled.Anchor>
      {<Styled.Clickable onClick={toggleState}>{anchor}</Styled.Clickable>}
      {state !== State.closed && (
        <>
          <Styled.Dropdown $closing={state === State.closing}>
            {children}
          </Styled.Dropdown>
          {state === State.open && <Styled.CancelArea onClick={toggleState} />}
        </>
      )}
    </Styled.Anchor>
  );
};

export default React.memo(Dropdown);
