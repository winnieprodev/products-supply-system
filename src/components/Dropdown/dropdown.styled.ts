import styled, { css, keyframes } from "styled-components";

export const Anchor = styled.div`
  position: relative;
  display: inline-block;
  z-index: +1;
`;

export const Clickable = styled.div`
  cursor: pointer;
`;

const openDropdownAnimation = keyframes`
    0% { 
        transform: scale(0.1); 
        opacity: 0.1; 
    }
    100% {
        transform: scale(1);
        opacity: 1;
    }
`;

const closedDropdownAnimation = keyframes`
    0% {
        transform: scale(1);
        opacity: 1;
    }
    100% { 
        transform: scale(0.1); 
        opacity: 0; 
    }
`;
interface DropdownProps {
  $closing: boolean;
}

export const CLOSE_ANIMATION_TIME = 300;

/* TODO: Make Dropdown slightly more re-usable with those fixed "top" / "right" values, 
   if it will be ever needed */
export const Dropdown = styled.div<DropdownProps>`
  position: absolute;
  top: 25px;
  right: -10px;
  background: white;
  padding: 16px 24px;
  border-radius: 6px;
  transform-origin: top right;
  animation: ${openDropdownAnimation} ease 0.25s;
  ${({ $closing }) =>
    $closing &&
    css`
      animation: ${closedDropdownAnimation} ease ${CLOSE_ANIMATION_TIME}ms;
      animation-fill-mode: forwards;
    `}
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.15), 0 4px 10px rgba(0, 0, 0, 0.15);

  &::before {
    pointer-events: none;
    display: block;
    content: "";
    position: absolute;
    top: -20px;
    right: 10px;
    height: 0;
    width: 0;
    border: 12px solid transparent;
    border-bottom-color: white;
  }
`;

export const CancelArea = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(255, 255, 255, 0.2);
  z-index: -1;
`;
