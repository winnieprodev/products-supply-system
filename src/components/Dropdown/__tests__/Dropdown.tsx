import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

import Dropdown from "../Dropdown";

describe("Dropdown", () => {
  describe("Snapshot tests", () => {
    test("it matches the snapshot", () => {
      const { container } = render(
        <BrowserRouter>
          <Dropdown anchor={<div>anchor</div>}>test</Dropdown>
        </BrowserRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
