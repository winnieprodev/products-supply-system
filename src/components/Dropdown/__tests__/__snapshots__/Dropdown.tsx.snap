// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Snapshot tests it matches the snapshot 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  z-index: +1;
}

.c1 {
  cursor: pointer;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div>
      anchor
    </div>
  </div>
</div>
`;
